misc/server/src/main.rs

```diff
@@ -77,35 +71,25 @@ async fn main() -> Result<(), Box<dyn Error>> {
 
         (peer_id, keypair)
     };
-
-    let transport = {
-        let tcp_transport =
-            tcp::tokio::Transport::new(tcp::Config::new().port_reuse(true).nodelay(true))
-                .upgrade(upgrade::Version::V1)
-                .authenticate(noise::Config::new(&local_keypair)?)
-                .multiplex(yamux::Config::default())
-                .timeout(Duration::from_secs(20));
-
-        let quic_transport = quic::tokio::Transport::new(quic::Config::new(&local_keypair));
-
-        dns::tokio::Transport::system(libp2p::core::transport::OrTransport::new(
-            quic_transport,
-            tcp_transport,
-        ))?
-        .map(|either_output, _| match either_output {
-            Either::Left((peer_id, muxer)) => (peer_id, StreamMuxerBox::new(muxer)),
-            Either::Right((peer_id, muxer)) => (peer_id, StreamMuxerBox::new(muxer)),
+    info!("Local peer id: {local_peer_id}");
+
+    let mut swarm = libp2p::SwarmBuilder::with_existing_identity(local_keypair)
+        .with_tokio()
+        .with_tcp(
+            tcp::Config::default().port_reuse(true).nodelay(true),
+            noise::Config::new,
+            yamux::Config::default,
+        )?
+        .with_quic_config(|key| {
```

Comment:
Would `with_custom_quic` perhaps be a better name? At the moment we have `with_quic` and `with_quic_config` and `with_quic_config` implies `with_quic` but I think technically, that it is not super obvious.

Also, we could change this closure to already give you an instance of `quic::Config`! Then the user would a) not have to name it and b) for builder-style configs, they can do them as one-liners:

```rust
	.with_quic_config(|cfg| cfg.with_draft_29().with_xyz())
```

Reply:
I find `with_quic_config` more intuitive. While `with_quic_config` implies that one can configure a `libp2p_quic::Config`, `with_custom_quic` does not imply what one can customize.

Do you worry that users think that `with_quic_config` only constructs the `libp2p_quic::Config` but not the actual `Transport`?

4743616ec makes `with_quic_config` take a closure.

Note that `libp2p_quic::Config` does not provide methods that can be chained. In case that is something worth doing, we can still do so in a follow-up pull request.

Reply:
> Do you worry that users think that `with_quic_config` only constructs the `libp2p_quic::Config` but not the actual `Transport`?

Not really, because there is no alternative function. I just figured the naming could be perceived as being a bit off :)